_posts/grant_guo/2018-03-22-kf-etl.md

```diff
@@ -0,0 +1,349 @@
+
+### Immutability
+
+Scala encourages developers to define immutable data structures and variables. Immutability helps keep control of the application states, avoid unexpected values, and avoid concurrent state issues. Collection types defined in the package `scala.collection.immutable` are immutable. Some functions in those types return a new instance of the same type instead of altering the collection in-place. This feature makes the method chaining possible. For example:
+```
+List("Jan", "Feb", "Mar", ..)
+    .filter(m => {???})
+    .map(m => {???})
+    .collect(m => {???})
+```
+
+Another interesting and  powerful language construct is `case classes` as the following definition:
+```
+case class Address(number:Int, street:String, city:String, province:String)
+```
+Scala compiler will add some syntactic conveniences to the class:
+* A factory method with the name of the class. This means that, for example, you can write `Address(661, "University Ave", "Toronto", "ON")` to construct a Address object. Under the hood, the `apply(...)` method generated in the companion object makes this magic possible.
+* All arguments in the parameter list of a case class implicitly get a `val` prefix
+* The "natural" implementations of methods `toString`, `hashCode`, and `equals`
+* A `copy` method, which is useful for making a new instance of the class that is the same as another one except that one or two attributes are different. For example:
+```
+val addr = Address(661, "University Ave", "Toronto", "ON")
+val new_addr = addr.copy(province = "Ontario")
+```
+
+### Higher-order functions
+
+In Scala, a function can be defined in the standard way:
+```def plusOne(x:Int):Int = {x + 1}```
+or as a functional literal:
+```val func = (x:Int) => x + 1```
+
+Basicaly, a higher-order function can have one the three forms:
+* one or more of its parameters is a function, and it returns some value
+* it returns a function, but none of its parameters are a function
+* both of the above
+
+Functions like `map`, `filter` or `flatMap` defined in collection types are the first type. In the section [Implementation Details]() we will examine how the ETL uses the other two types, especially function composition to make function chaining possible.
+
+### Pattern Matching
+
+Pattern Matching in Scala is amazingly powerful. Here are some examples:
+* wildcard, varable, constructor patterns:
+```
+expr match {
+    case Address(number, street, _, _) => println(s"${number}, ${street}")
+    ...
+}
+```
+* constant patterns:
+```
+expr match {
+    case 10 => "ten"
+    case false => "false"
+    case "hello" => "hi!"
+    case Nil => "the empty list"
+    case _ => "something else"
+}
+```
+* sequence patterns:
+```
+list match {
+    case List(0, _, _) => println("the list starts with 0")
+    case _ => 
+}
+```
+* typed patterns
+```
+expr match {
+    case s: String => s.length
+    case m: Map[_, _] => m.size
+    case _ => -1
+}
+```
+* variable binding:
```

Comment:
From that point on, code blocks are not in the correct place (i.e. text appear in code blocks and the other way around).